renjin/org/renjin/cran/stringi/stringi.java

```diff
@@ -164,12 +239,238 @@ public static void R_init_stringi(DllInfo dll) {
+    final StringVector[] result = new StringVector[length];
+    final ExpressionVector strings = __ensure_length(length, str);
+    final ExpressionVector patterns = __ensure_length(length, pattern);
+    final ExpressionVector ns = __ensure_length(length, n);
+    final ExpressionVector omits = __ensure_length(length, omit_empty);
+
+    String lastPattern = null;
+    UnicodeSet matcher = null;
+    for (int i = 0; i < length; i++) {
+      if (strings.isElementNA(i) || patterns.isElementNA(i) || ns.isElementNA(i)) {
+        result[i] = StringVector.valueOf(StringVector.NA);
+      } else {
+        final boolean omit = omits.getElementAsLogical(i).toBooleanStrict();
+        int depth = ns.getElementAsInt(i);
+        if (depth == 0) {
+          result[i] = StringVector.EMPTY;
+        } else {
+          if (depth >= Integer.MAX_VALUE - 1) {
+            throw new IllegalArgumentException("argument `" + depth + "`: value too large");
+          } else if (depth < 0) {
+            depth = Integer.MAX_VALUE;
+          } else if (only_tokens) {
+            depth++; // we need to do one split ahead here
+          }
+          final LinkedList<String> fields = new LinkedList<String>();
+          final String element = strings.getElementAsString(i);
+          final String separatorPattern = __normalize_binary_properties(patterns.getElementAsString(i));
+          if (!separatorPattern.equals(lastPattern)) {
+            lastPattern = separatorPattern;
+            matcher = new UnicodeSet(separatorPattern);
+          }
+          int previousStart = 0;
+          for (int k = 0; previousStart < element.length() && k < depth;) {
+            final int beginIndex = matcher.span(element, previousStart, UnicodeSet.SpanCondition.NOT_CONTAINED);
+            final int endIndex = matcher.span(element, beginIndex, UnicodeSet.SpanCondition.CONTAINED);
+            if (omit && previousStart == beginIndex) {
+              // don't start any new field
+            } else { // start a new field here
+              final String field = element.substring(previousStart, beginIndex);
+              fields.add(("".equals(field) && omits.isElementNA(i)) ? StringVector.NA : field);
+              ++k; // another field
+            }
+            previousStart = endIndex;
+          }
+          if (!(omit && previousStart == element.length())) {
+            final String field = element.substring(previousStart, element.length());
+            fields.add(("".equals(field) && omits.isElementNA(i)) ? StringVector.NA : field);
+          }
+          if (only_tokens && depth < Integer.MAX_VALUE) {
+            depth--; // one split ahead could have been made, see above
+            while (fields.size() > depth) {
+              fields.removeLast(); // get rid of the remainder
+            }
+          }
+          result[i] = new StringArrayVector(fields);
+        }
+      }
+    }
+
+    final ExpressionVector resultSexp = new ExpressionVector(result);
```

Comment:
I don't think you mean `ExpressionVector` here. `ExpressionVector` is a Java type for the R type "expression", which is a list of unevaluated expressions, like `a+1; foo(); c`. You probably want `ListVector`, which corresponds to the R type "list" ?